src/main/scala/loamstream/model/execute/RxExecuter.scala

```diff
@@ -54,58 +55,111 @@ final case class RxExecuter(
     
-    //An observable stream of "chunks" of runnable jobs, with each chunk represented as a Seq.
-    //Jobs are buffered up until the amount of time indicated by 'windowLength' elapses, or 'runner.maxNumJobs'
-    //are collected.  When that happens, the buffered "chunk" of jobs is emitted.
-    val chunks: Observable[Seq[JobRun]] = runnables.tumblingBuffer(windowLength, runner.maxNumJobs, ioScheduler)
+    val executionState = TimeUtils.time(s"Initializing execution state with ${numJobs} jobs", debug(_)) {
+      ExecutionState.initialFor(executable, maxRunsPerJob)
+    }
+
+    val ioScheduler: Scheduler = IOScheduler()
     
-    val chunkResults: Observable[Map[LJob, Execution]] = for {
-      //NB: .toSet is important: jobs in a chunk should be distinct, 
-      //so they're not run more than once before transitioning to a terminal state.
-      jobs <- chunks.map(_.toSet)
-      //NB: Filter out jobs from this chunk that finished when run as part of another chunk, so we don't run them
-      //more times than necessary.  This helps in the face of job-restarting, since we can't call `distinct()` 
-      //on `runnables` and declare victory like we did before, since that would filter out restarting jobs that 
-      //already ran 
-      (finishedJobs, notFinishedJobs) = jobs.partition(_.status.isTerminal)
-      if notFinishedJobs.nonEmpty
-      (jobsToMaybeRun, skippedJobs) = notFinishedJobs.map(_.job).partition(jobFilter.shouldRun)
-      (jobsToCancel, jobsToRun) = jobsToMaybeRun.partition(jobCanceler.shouldCancel)
-      _ = handleSkippedJobs(skippedJobs)
-      cancelledJobsMap = cancelJobs(jobsToCancel)
-      _ = record(jobOracle, cancelledJobsMap)
-      skippedResultMap = toSkippedResultMap(skippedJobs)
-      executionTupleOpt <- runJobs(jobsToRun, jobOracle)
-      _ = record(jobOracle, executionTupleOpt)
-      executionMap = cancelledJobsMap ++ executionTupleOpt
-      _ = logFinishedJobs(executionMap)
-    } yield {
-      executionMap ++ skippedResultMap
+    val chunkResults: Observable[Map[LJob, Execution]] = {
+      //Note onBackpressureDrop(), in case runEligibleJobs takes too long (or the polling window is too short)
+      val ticks = Observable.interval(windowLength, ioScheduler).onBackpressureDrop
+      
+      def runJobs() = runEligibleJobs(executionState, jobOracle)
+      
+      def isFinished = executionState.isFinished
+      
+      ticks.flatMap(_ => runJobs).takeUntil(_ => isFinished)
     }
-    //NB: We no longer stop on the first failure, but run each sub-tree of jobs as far as possible.
     
     val futureMergedResults = chunkResults.foldLeft(emptyExecutionMap)(_ ++ _).firstAsFuture
 
     Await.result(futureMergedResults, timeout)
   }
   
+  private def finish(executionState: ExecutionState)(results: Map[LJob, Execution]): Map[LJob, Execution] = {
+    def msg = {
+      val howMany: Int = scala.math.min(50, results.size) //scalastyle:ignore magic.number
```

Comment:
Just noting that `.take` won't crash if too many, so you could do the `take` first and then use `.size` of the resulting list to know how many it actually took. 6 vs. 1/2 dozen.

Reply:
You're right, thanks.